src/encode.ts

```diff
@@ -0,0 +1,477 @@
+							(async () => {
+								let iterator = (promise as AsyncIterable<unknown>)[
+									Symbol.asyncIterator
+								]();
+
+								let result: IteratorResult<unknown>;
+								do {
+									result = await iterator.next();
+									if (!result.done) {
+										controller.enqueue(`${id}${STR_SUCCESS}`);
+										encode(result.value);
+									}
+								} while (!result.done);
+							})()
+								.then(() => {
+									controller.enqueue(`${id}\n`);
+								})
+								.catch((error) => {
+									controller.enqueue(`${id}${STR_FAILURE}`);
+									encode(error);
+								})
+								.finally(() => {
+									wg.done();
+								});
+						}
+					}
+				},
+			};
+
+			try {
+				encode(value);
+
+				do {
+					await wg.wait();
+				} while (wg.p > 0);
+
+				controller.close();
+			} catch (error) {
+				controller.error(error);
+			}
+		},
+	});
+}
+
+type EncodeFrameObj =
+	| { type: 1; prefix: string; value: unknown }
+	| { type: 2; prefix: string; value: undefined };
+
+class EncodeFrame {
+	public type: number;
+	public prefix: string;
+	public value: unknown;
+	constructor(type: number, prefix: string, value: unknown) {
+		this.type = type;
+		this.prefix = prefix;
+		this.value = value;
+	}
+}
+
+export function encodeSync(
+	value: unknown,
+	chunks: { push(...chunk: string[]): void },
+	refs: WeakMap<object, number>,
+	promises: WeakMap<object, number>,
+	asyncFrames: { push(frame: AsyncFrame): void },
+	counters: { refId: number; promiseId: number },
+	plugins: EncodePlugin[],
+	redactErrors: boolean,
+) {
+	let encodeStack: EncodeFrameObj[] = [
+		new EncodeFrame(1, "", value) as EncodeFrameObj,
+	];
+	let frame: EncodeFrameObj | undefined;
+
+	encodeLoop: while ((frame = encodeStack.pop()) !== undefined) {
+		if (frame.type === 2) {
+			chunks.push(frame.prefix);
+			continue;
+		}
+
+		let { prefix, value } = frame;
+		chunks.push(prefix);
+
+		if (value === undefined) {
+			chunks.push(STR_UNDEFINED);
+			continue;
+		}
+		if (value === null) {
+			chunks.push(STR_NULL);
+			continue;
+		}
+		if (value === true) {
+			chunks.push(STR_TRUE);
+			continue;
+		}
+		if (value === false) {
+			chunks.push(STR_FALSE);
+			continue;
+		}
+
+		const typeOfValue = typeof value;
+		if (typeOfValue === "object") {
+			if (
+				value instanceof Promise ||
+				typeof (value as PromiseLike<unknown>).then === "function"
+			) {
+				let existingId = promises.get(value);
+				if (existingId !== undefined) {
+					chunks.push(STR_PROMISE, existingId.toString());
+					continue;
+				}
+
+				let promiseId = counters.promiseId++;
+				promises.set(value, promiseId);
+				chunks.push(STR_PROMISE, promiseId.toString());
+				asyncFrames.push([1, promiseId, value as PromiseLike<unknown>]);
+				continue;
+			}
+
+			if (value instanceof ReadableStream) {
+				let existingId = promises.get(value);
+				if (existingId !== undefined) {
+					chunks.push(STR_READABLE_STREAM, existingId.toString());
+					continue;
+				}
```

Comment:
Couldn't this also be handled by the `STR_REFERENCE_SYMBOL` later? That would probably save a bunch of bundle size (same for the other types)

Reply:
It could but would make it a lot less readable unless I'm missing something. The problem with combining the cases is that ReadableStreams need to be handled differently meaning that we'd have to introduce more nested conditions making hot paths less predictable.